compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp

```diff
@@ -568,6 +569,138 @@ void transform_dialect::CopyTensorOperandOp::getEffects(
   transform::modifiesPayload(effects);
 }
 
+//===---------------------------------------------------------------------===//
+// FlattenForallOp
+//===---------------------------------------------------------------------===//
+
+template <typename MappingTy>
+bool isAscendingRelativeMapping(ArrayRef<Attribute> mapping) {
+  auto start = cast<MappingTy>(*mapping.begin());
+  if (start.getMappingId() !=
+      static_cast<int64_t>(gpu::MappingId::LinearDim0)) {
+    return false;
+  }
+  int64_t prev = start.getMappingId();
+  for (auto id : llvm::drop_begin(mapping)) {
+    int64_t nextId = cast<MappingTy>(id).getMappingId();
+    if (nextId != prev + 1) {
+      return false;
+    }
+    prev = nextId;
+  }
+  return true;
+}
+
+FailureOr<scf::ForallOp> flattenForallOp(RewriterBase &rewriter,
+                                         scf::ForallOp forallOp) {
+  if (!forallOp.getMapping().has_value())
+    return forallOp->emitError("mapping must be present");
+  SmallVector<Attribute> mapping =
+      llvm::to_vector(forallOp.getMapping()->getValue());
+  if (!(llvm::all_of(mapping, llvm::IsaPred<gpu::GPUThreadMappingAttr>) ||
+        llvm::all_of(mapping, llvm::IsaPred<gpu::GPUWarpMappingAttr>))) {
+    return forallOp->emitError("mapping must be #gpu.thread or #gpu.warp");
+  }
+
+  if (forallOp.getRank() == 1) {
+    return forallOp;
+  }
+
+  bool isThreadMapping = isa<gpu::GPUThreadMappingAttr>(*mapping.begin());
+
+  if (isThreadMapping) {
+    if (!isAscendingRelativeMapping<gpu::GPUThreadMappingAttr>(mapping)) {
+      return forallOp->emitError(
+          "mapping must be descending linear thread ids");
+    }
+  } else if (!isAscendingRelativeMapping<gpu::GPUWarpMappingAttr>(mapping)) {
+    return forallOp->emitError("mapping must be descending linear warp ids");
+  }
+
+  auto isAll = [](ArrayRef<int64_t> array, int64_t cmp) {
+    return llvm::all_of(array, [cmp](int64_t x) { return x == cmp; });
+  };
+
+  if (!isAll(forallOp.getStaticStep(), 1) ||
+      !isAll(forallOp.getStaticLowerBound(), 0)) {
+    return failure();
+    return forallOp->emitError("");
+  }
+
+  MLIRContext *context = rewriter.getContext();
+  Location loc = forallOp.getLoc();
+
+  // Step 1. Construct the new mapping attribute as a single linear dim of the
+  // original type.
+  Attribute flatMapping;
+  if (isThreadMapping) {
+    flatMapping =
+        gpu::GPUThreadMappingAttr::get(context, gpu::MappingId::LinearDim0);
+  } else {
+    flatMapping =
+        gpu::GPUWarpMappingAttr::get(context, gpu::MappingId::LinearDim0);
+  }
```

Comment:
Can probably save some lines and nesting here
```suggestion
  Attribute flatMapping = isThreadMapping ?
    gpu::GPUThreadMappingAttr::get(context, gpu::MappingId::LinearDim0) :
    gpu::GPUWarpMappingAttr::get(context, gpu::MappingId::LinearDim0);
```

Reply:
I tried this and I couldn't get clang to stop yelling at me. It doesn't like the implicit cast to `Attribute` on the ternary operator results. Maybe someone better at C++ could figure it out...